File: demos/01-essentials/01-ai-foundry/05-rag/rag-azure-py/search_index_manager.py
from azure.search.documents.indexes import SearchIndexClient
from azure.search.documents.indexes.models import (
    SearchIndex,
    SearchField,
    SearchFieldDataType,
    SimpleField,
    SearchableField,
    VectorSearch,
    VectorSearchProfile,
    HnswAlgorithmConfiguration,
    ExhaustiveKnnAlgorithmConfiguration,
    SemanticConfiguration,
    SemanticPrioritizedFields,
    SemanticField,
    SemanticSearch,
)
from azure.core.exceptions import ResourceNotFoundError
from typing import Dict


class SearchIndexManager:
    def __init__(self, search_index_client: SearchIndexClient, index_name: str):
        self.search_index_client = search_index_client
        self.index_name = index_name

    def create_search_index(self, embeddings_model: str) -> bool:
        print(f"🚀 Creating search index with manual embeddings using Azure AI Project")
        print(f"🤖 Embeddings model: {embeddings_model}")
        
        vector_search = VectorSearch(
            algorithms=[
                HnswAlgorithmConfiguration(name="insurance-algorithm", kind="hnsw"),
                ExhaustiveKnnAlgorithmConfiguration(name="insurance-eknn", kind="exhaustiveKnn")
            ],
            profiles=[
                VectorSearchProfile(
                    name="insurance-profile",
                    algorithm_configuration_name="insurance-algorithm"
                )
            ]
        )
        
        semantic_config = SemanticConfiguration(
            name="insurance-semantic",
            prioritized_fields=SemanticPrioritizedFields(
                title_field=SemanticField(field_name="title"),
                content_fields=[SemanticField(field_name="content")],
                keywords_fields=[
                    SemanticField(field_name="category"),
                    SemanticField(field_name="file_name")
                ]
            )
        )
        
        semantic_search = SemanticSearch(
            configurations=[semantic_config]
        )
        
        fields = [
            SimpleField(name="id", type=SearchFieldDataType.String, key=True),
            SearchableField(name="title", type=SearchFieldDataType.String),
            SearchableField(name="content", type=SearchFieldDataType.String),
            SearchableField(name="category", type=SearchFieldDataType.String, filterable=True, facetable=True),
            SearchableField(name="file_name", type=SearchFieldDataType.String, filterable=True),
            SimpleField(name="file_type", type=SearchFieldDataType.String, filterable=True),
            SimpleField(name="chunk_id", type=SearchFieldDataType.Int32),
            SimpleField(name="chunk_count", type=SearchFieldDataType.Int32),
            SimpleField(name="original_length", type=SearchFieldDataType.Int32),
            SimpleField(name="chunk_length", type=SearchFieldDataType.Int32),
            SimpleField(name="processing_date", type=SearchFieldDataType.DateTimeOffset),
            SearchField(
                name="content_vector",
                type=SearchFieldDataType.Collection(SearchFieldDataType.Single),
                searchable=True,
                vector_search_dimensions=1536,
                vector_search_profile_name="insurance-profile"
            )
        ]
        
        index = SearchIndex(
            name=self.index_name,
            fields=fields,
            vector_search=vector_search,
            semantic_search=semantic_search
        )
        
        result = self.search_index_client.create_or_update_index(index)
        print(f"✅ Search index '{self.index_name}' created successfully!")
        print(f"📋 Index fields: {len(result.fields)}")
        print(f"🔍 Vector search enabled: {bool(result.vector_search)}")
        print(f"🧠 Semantic search enabled: {bool(result.semantic_search)}")
        print(f"📝 Embeddings will be generated manually using Azure AI Project")
        
        return True
    
    def delete_index_if_exists(self) -> bool:
        try:
            self.search_index_client.delete_index(self.index_name)
            print(f"✅ Deleted existing index: {self.index_name}")
            return True
        except ResourceNotFoundError:
            print(f"ℹ️ Index {self.index_name} doesn't exist - will create new")
            return True
    
    def get_index_stats(self) -> Dict:
        try:
            index = self.search_index_client.get_index(self.index_name)
            stats = self.search_index_client.get_index_statistics(self.index_name)
            
            if hasattr(stats, 'document_count'):
                return {
                    "name": index.name,
                    "field_count": len(index.fields),
                    "document_count": stats.document_count,
                    "storage_size": stats.storage_size,
                    "vector_index_size": getattr(stats, 'vector_index_size', 0)
                }
            else:
                return {
                    "name": index.name,
                    "field_count": len(index.fields),
                    "document_count": stats.get('document_count', 0),
                    "storage_size": stats.get('storage_size', 0),
                    "vector_index_size": stats.get('vector_index_size', 0)
                }
        except Exception as e:
            print(f"❌ Error getting index stats: {e}")
            return {}

File: demos/01-essentials/02-agentic-ai/02-rag/rag-azure-cs/SearchIndexManager.cs
using Azure;
using Azure.Search.Documents.Indexes;
using Azure.Search.Documents.Indexes.Models;

namespace RagAzure;

public class SearchIndexManager(SearchIndexClient searchIndexClient, string indexName)
{

    public async Task<bool> CreateSearchIndexAsync(string embeddingsModel)
    {
        Console.WriteLine("\n## 3. Create Azure AI Search Index with Integrated Vectorization");
        Console.WriteLine($"🚀 Creating search index with manual embeddings using Azure AI Project");
        Console.WriteLine($"🤖 Embeddings model: {embeddingsModel}");

        var vectorSearch = new VectorSearch();
        vectorSearch.Algorithms.Add(new HnswAlgorithmConfiguration("insurance-algorithm"));
        vectorSearch.Algorithms.Add(new ExhaustiveKnnAlgorithmConfiguration("insurance-eknn"));
        vectorSearch.Profiles.Add(new("insurance-profile", "insurance-algorithm"));

        var semanticConfig = new SemanticConfiguration(
            "insurance-semantic",
            new SemanticPrioritizedFields
            {
                TitleField = new SemanticField("title"),
                ContentFields =
                {
                    new SemanticField("content")
                },
                KeywordsFields =
                {
                    new SemanticField("category"),
                    new SemanticField("file_name")
                }
            });

        var semanticSearch = new SemanticSearch();
        semanticSearch.Configurations.Add(semanticConfig);

        List<SearchField> fields =
        [
            new SimpleField("id", SearchFieldDataType.String) { IsKey = true },
            new SearchableField("title") { IsFilterable = false, IsSortable = false },
            new SearchableField("content") { IsFilterable = false, IsSortable = false },
            new SearchableField("category") { IsFilterable = true, IsFacetable = true },
            new SearchableField("file_name") { IsFilterable = true },
            new SimpleField("file_type", SearchFieldDataType.String) { IsFilterable = true },
            new SimpleField("chunk_id", SearchFieldDataType.Int32),
            new SimpleField("chunk_count", SearchFieldDataType.Int32),
            new SimpleField("original_length", SearchFieldDataType.Int32),
            new SimpleField("chunk_length", SearchFieldDataType.Int32),
            new SimpleField("processing_date", SearchFieldDataType.DateTimeOffset),
            new SearchField("content_vector", SearchFieldDataType.Collection(SearchFieldDataType.Single))
            {
                IsSearchable = true,
                VectorSearchDimensions = 1536,
                VectorSearchProfileName = "insurance-profile"
            }
        ];

        var index = new SearchIndex(indexName)
        {
            Fields = fields,
            VectorSearch = vectorSearch,
            SemanticSearch = semanticSearch
        };

        var result = await searchIndexClient.CreateOrUpdateIndexAsync(index);

        Console.WriteLine($"✅ Search index '{indexName}' created successfully!");
        Console.WriteLine($"📋 Index fields: {result.Value.Fields.Count}");
        Console.WriteLine($"🔍 Vector search enabled: {result.Value.VectorSearch != null}");
        Console.WriteLine($"🧠 Semantic search enabled: {result.Value.SemanticSearch != null}");
        Console.WriteLine($"📝 Embeddings will be generated manually using Azure AI Project");

        return true;
    }

    public async Task<bool> DeleteIndexIfExistsAsync()
    {
        try
        {
            await searchIndexClient.DeleteIndexAsync(indexName);
            Console.WriteLine($"✅ Deleted existing index: {indexName}");
            return true;
        }
        catch (RequestFailedException ex) when (ex.Status == 404)
        {
            Console.WriteLine($"ℹ️ Index {indexName} doesn't exist - will create new");
            return true;
        }
    }

    public async Task<Dictionary<string, object>> GetIndexStatsAsync()
    {
        try
        {
            var index = await searchIndexClient.GetIndexAsync(indexName);
            var stats = await searchIndexClient.GetIndexStatisticsAsync(indexName);

            return new()
            {
                ["name"] = index.Value.Name,
                ["field_count"] = index.Value.Fields.Count,
                ["document_count"] = stats.Value.DocumentCount,
                ["storage_size"] = stats.Value.StorageSize
            };
        }
        catch (Exception ex)
        {
            Console.WriteLine($"❌ Error getting index stats: {ex.Message}");
            return [];
        }
    }
}
